### PR TITLE
[Performance]Qwen3-Omni performance optimization

### PR DESCRIPTION
**Pull request by amy-why-3459**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose

1. Code2Wav does not require _update_request_states during the idle phase.
2. Only pass thinker_hidden_states during the prefill phase.
3. Convert the save_async operation to asynchronous.
4. Modify the method of _update_request_information

#1377 

## Test Plan

```
vllm serve Qwen/Qwen3-Omni-30B-A3B-Instruct 
     --omni 
     --port 8091 
     --stage-configs-path /vllm_omni/model_executor/stage_configs/qwen3_omni_moe_async_chunk.yaml
```

## Test Result

```
vllm bench serve \
    --omni \
  --dataset-name random \
  --port 28889 \
  --max-concurrency 1 \
  --model Qwen/Qwen3-Omni-30B-A3B-Instruct \
  --endpoint /v1/chat/completions \
  --backend openai-chat-omni \
  --num-prompts 1 \
  --random-input-len 2500 \
  --ignore-eos \
  --percentile-metrics ttft,tpot,itl,e2el,audio_ttfp,audio_rtf \
  --random-output-len 900 \
  --extra_body '{"modalities": ["text", "audio"]}'
```
```
============ Serving Benchmark Result ============
Successful requests:                     1
Failed requests:                         0
Maximum request concurrency:             1
Benchmark duration (s):                  54.17
Request throughput (req/s):              0.02
Peak concurrent requests:                1.00
----------------End-to-end Latency----------------
Mean E2EL (ms):                          54166.65
Median E2EL (ms):                        54166.65
P99 E2EL (ms):                           54166.65
================== Text Result ===================
Total input tokens:                      2500
Total generated tokens:                  900
Output token throughput (tok/s):         16.61
Peak output token throughput (tok/s):    85.00
Peak concurrent requests:                1.00
Total Token throughput (tok/s):          62.77
---------------Time to First Token----------------
Mean TTFT (ms):                          308.13
Median TTFT (ms):                        308.13
P99 TTFT (ms):                           308.13
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          13.28
Median TPOT (ms):                        13.28
P99 TPOT (ms):                           13.28
---------------Inter-token Latency----------------
Mean ITL (ms):                           13.28
Median ITL (ms):                         12.24
P99 ITL (ms):                            133.63
================== Audio Result ==================
Total audio duration generated(s):       211.66
Total audio frames generated:            5079900
Audio throughput(audio duration/s):      3.91
---------------Time to First Packet---------------
Mean AUDIO_TTFP (ms):                    1370.63
Median AUDIO_TTFP (ms):                  1370.63
P99 AUDIO_TTFP (ms):                     1370.63
-----------------Real Time Factor-----------------
Mean AUDIO_RTF:                          0.26
Median AUDIO_RTF:                        0.26
P99 AUDIO_RTF:                           0.26
==================================================
```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please providing the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please pasting the results comparison before and after, or e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/distributed/omni_connectors/transfer_adapter/chunk_transfer_adapter.py [hunk 1]
```diff
@@ -109,43 +109,18 @@ def save_async(self, pooling_output: torch.Tensor | None = None, request: Reques
...
+                "pooling_output": pooling_output,
+                "request": request,
+            }
             if request_id not in self._pending_save_reqs:
                 self._pending_save_reqs[request_id] = deque()
             self._pending_save_reqs[request_id].append(task)
+            self.put_req_chunk[request_id] += 1
 
     def _poll_single_request(self, req_id: str):
         stage_id = self.connector.stage_id
```
Comment (lishunyang12) [inline vllm_omni/distributed/omni_connectors/transfer_adapter/chunk_transfer_adapter.py:117]: I was tracing through the change and I think this is a nice improvement -- `save_async` now only captures the raw `pooling_output` and `request` under the lock, and the expensive `custom_process_next_stage_input_func` call moves to `_send_single_request`.

However, I noticed that in the original code, `save_async` would `return` early if `payload_data` was falsy (i.e., the processor returned nothing). With the new code, we always enqueue the task, and the early return happens later in `_send_single_request`. This means the `chunk_id` is incremented and a task is queued even when the processor will ultimately produce no payload. Could this leave gaps in the chunk_id sequence that downstream consumers might not expect?

File: vllm_omni/model_executor/stage_input_processors/qwen3_omni.py [hunk 2]
```diff
@@ -134,10 +134,12 @@ def thinker2talker_async_chunk(
...
             "thinker_embeddings": pooling_output.get("0").detach().cpu(),
-            "thinker_hidden_states": pooling_output.get("24").detach().cpu(),
-            "thinker_sequences": output_token_ids,
             "finished": torch.tensor(request.is_finished(), dtype=torch.bool),
         }
+
+        if not output_token_ids:
+            # When prefilling a chunked thinker, thinker_hidden_states needs to be updated.
+            talker_additional_info["thinker_hidden_states"] = pooling_output.get("24").detach().cpu()
     return talker_additional_info
```
Comment (lishunyang12) [inline vllm_omni/model_executor/stage_input_processors/qwen3_omni.py:144]: I wanted to make sure I understand this change correctly. Previously, `thinker_hidden_states` and `thinker_sequences` were sent on every chunk. Now `thinker_hidden_states` is only sent when `output_token_ids` is empty (prefill), and `thinker_sequences` is removed entirely.

Does this mean the talker no longer needs `thinker_sequences` at all? And for `thinker_hidden_states`, the talker only needs it once during prefill and can reuse it across subsequent chunks? A brief comment explaining this would help future readers understand the performance rationale.

File: vllm_omni/worker/gpu_model_runner.py [hunk 3]
```diff
@@ -911,16 +911,28 @@ def _collect_additional_information_for_prefill(
...
+        """Update per-request additional_information stored in request state."""
+        req_state = self.requests.get(request_id)
+        if req_state is None:
+            return
+        if not hasattr(req_state, "additional_information_cpu"):
+            return
+
+        info_dict = getattr(req_state, "additional_information_cpu")
+        if isinstance(payload_info, dict):
+            info_dict.update(payload_info)
```
Comment (lishunyang12) [inline vllm_omni/worker/gpu_model_runner.py:935]: I see that `_update_request_information` replaces calls to `_merge_additional_information_update`. I was curious about the difference -- does `_merge_additional_information_update` do a deeper merge (e.g., nested dict merging), while this new method does a shallow `dict.update()`? If so, could there be cases where nested keys get overwritten instead of merged? If the payloads are always flat dicts, this is fine, but it would be helpful to add a comment noting this assumption.

File: vllm_omni/distributed/omni_connectors/transfer_adapter/chunk_transfer_adapter.py [hunk 4]
```diff
@@ -109,43 +109,18 @@ def save_async(self, pooling_output: torch.Tensor | None = None, request: Reques
...
-        }
-
         with self.lock:
+            request_id = request.external_req_id
+            chunk_id = self.put_req_chunk[request_id]
+            task = {
+                "chunk_id": chunk_id,
+                "pooling_output": pooling_output,
+                "request": request,
+            }
```
Comment (lishunyang12) [inline vllm_omni/distributed/omni_connectors/transfer_adapter/chunk_transfer_adapter.py]: The task now stores a live reference to `request`. When `_send_single_request` runs later, it reads `request.is_finished()`, `request.all_token_ids`, `request.prompt_token_ids`, etc. Since the request object is mutable, those values could have changed between enqueue time and send time. The old code processed everything eagerly in `save_async` so this was not an issue.

Is `_send_single_request` always called quickly enough that this does not matter in practice? Or should we snapshot the fields we need here?

File: vllm_omni/worker/gpu_generation_model_runner.py [hunk 5]
```diff
@@ -100,7 +100,7 @@ def execute_model(
             record_function_or_nullcontext("gpu_model_runner: preprocess"),
             self.synchronize_input_prep(),
         ):
-            if self.model_config.async_chunk:
+            if self.model_config.async_chunk and num_scheduled_tokens:
```
Comment (lishunyang12) [inline vllm_omni/worker/gpu_generation_model_runner.py:103]: What is the scenario where `async_chunk` is true but `num_scheduled_tokens` is 0? Is this guarding against a specific crash, or is it a perf skip? If the scheduler output carries metadata updates (e.g., finished status) but no tokens, would skipping `_update_request_states` cause staleness?

File: vllm_omni/worker/gpu_model_runner.py [hunk 6]
```diff
@@ -911,16 +911,28 @@ def _collect_additional_information_for_prefill(
                 start_offset = int(self.query_start_loc.cpu[req_index])
                 self.inputs_embeds[start_offset : start_offset + overlay_len].copy_(src)
 
+    def _update_request_information(self, request_id: str, payload_info: dict) -> None:
+        """Update per-request additional_information stored in request state."""
+        req_state = self.requests.get(request_id)
+        if req_state is None:
```
Comment (lishunyang12) [inline vllm_omni/worker/gpu_model_runner.py:930]: Nit: `hasattr` + `getattr` on the same attribute back to back -- you could just do:
```python
info_dict = getattr(req_state, "additional_information_cpu", None)
if info_dict is None:
    return
```

**Comment by hsliuustc0106** [comment]
@vllm-omni-reviewer

**Comment by hsliuustc0106** [comment]
This PR optimizes Qwen3-Omni performance by refactoring the async chunk transfer mechanism. Key improvements include moving expensive processing outside locks and optimizing data transfer between stages. However, there are important concerns raised in the review comments about race conditions with mutable request objects and the removal of thinker_sequences that need to be addressed before merging.

File: vllm_omni/distributed/omni_connectors/transfer_adapter/chunk_transfer_adapter.py [hunk 7]
```diff
@@ -109,47 +106,17 @@ def save_async(self, pooling_output: torch.Tensor | None = None, request: Reques
...
-        connector_put_key = f"{request_id}_{stage_id}_{chunk_id}"
-
         task = {
-            "stage_id": stage_id,
-            "next_stage_id": next_stage_id,
-            "put_key": connector_put_key,
-            "data": payload_data,
-            "request_id": request_id,
+            "pooling_output": pooling_output,
+            "request": request,
```
Comment (hsliuustc0106) [inline vllm_omni/distributed/omni_connectors/transfer_adapter/chunk_transfer_adapter.py:111]: As noted in the review, storing a live reference to the mutable `request` object could lead to race conditions. The request's state (is_finished, all_token_ids, etc.) may change between enqueue and send time. Consider capturing immutable snapshots of needed fields at enqueue time instead.

File: vllm_omni/distributed/omni_connectors/transfer_adapter/chunk_transfer_adapter.py [hunk 8]
```diff
@@ -215,10 +178,30 @@ def _update_request_payload(self, req_id: str, payload_data: dict[str, Any]) ->
...
+        request = task["request"]
+        is_finished = task["is_finished"]
+        stage_id = self.connector.stage_id
+        next_stage_id = stage_id + 1
+        request_id = request.external_req_id
+        chunk_id = self.put_req_chunk[request_id]
+        connector_put_key = f"{request_id}_{stage_id}_{chunk_id}"
+        # Process payload in main thread to avoid race conditions on request state
+        payload_data = None
+        if self.custom_process_next_stage_input_func:
```
Comment (hsliuustc0106) [inline vllm_omni/distributed/omni_connectors/transfer_adapter/chunk_transfer_adapter.py:199]: Good optimization - moving the expensive custom_process_next_stage_input_func call outside the lock in save_async reduces contention. This should improve throughput.

File: vllm_omni/model_executor/stage_input_processors/qwen3_omni.py [hunk 9]
```diff
@@ -134,10 +135,12 @@ def thinker2talker_async_chunk(
 
         talker_additional_info = {
             "thinker_embeddings": pooling_output.get("0").detach().cpu(),
-            "thinker_hidden_states": pooling_output.get("24").detach().cpu(),
-            "thinker_sequences": output_token_ids,
-            "finished": torch.tensor(request.is_finished(), dtype=torch.bool),
+            "finished": torch.tensor(is_finished, dtype=torch.bool),
         }
+
+        if not output_token_ids:
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/stage_input_processors/qwen3_omni.py:141]: As flagged by Codex, this emptiness check may be problematic. After _ensure_list(), output_token_ids is a list, so `not output_token_ids` checks for empty list. Verify this is the intended behavior vs checking for None.

File: vllm_omni/worker/gpu_model_runner.py [hunk 10]
```diff
@@ -911,16 +911,28 @@ def _collect_additional_information_for_prefill(
...
+            return
+        if not hasattr(req_state, "additional_information_cpu"):
```
Comment (hsliuustc0106) [inline vllm_omni/worker/gpu_model_runner.py]: As suggested in the review, you can simplify this with `info_dict = getattr(req_state, "additional_information_cpu", None)` and then check `if info_dict is None: return`. This avoids the redundant hasattr/getattr pattern.

**Comment by amy-why-3459** [comment]
@Shirley125 @R2-Y PTAL

File: vllm_omni/model_executor/stage_input_processors/qwen3_omni.py (hunk 2 above)
Comment (amy-why-3459) [reply vllm_omni/model_executor/stage_input_processors/qwen3_omni.py:144]: We have already assigned all_token_ids to thinker_sequences, so there is no need to add it again later.

File: vllm_omni/worker/gpu_model_runner.py (hunk 3 above)
Comment (amy-why-3459) [reply vllm_omni/worker/gpu_model_runner.py:935]: We found that `_merge_additional_information_update` was time-consuming with long inputs, so we replaced it with `_update_request_information`.

File: vllm_omni/distributed/omni_connectors/transfer_adapter/chunk_transfer_adapter.py (hunk 4 above)
Comment (amy-why-3459) [reply vllm_omni/distributed/omni_connectors/transfer_adapter/chunk_transfer_adapter.py]: The `request.is_finished()` function has been added to the task. `prompt_token_ids` and `all_token_ids` are cumulative, so updates to them are fine.

File: vllm_omni/worker/gpu_generation_model_runner.py (hunk 5 above)
Comment (amy-why-3459) [reply vllm_omni/worker/gpu_generation_model_runner.py:103]: This is for performance optimization. When num_scheduled_tokens is 0, it means the model is idling and _update_request_states is unnecessary.

**Comment by amy-why-3459** [comment]
@lishunyang12 Please check again.

**Comment by amy-why-3459** [comment]
@Sy0307 PTAL

File: vllm_omni/distributed/omni_connectors/transfer_adapter/base.py [hunk 11]
```diff
@@ -48,37 +48,30 @@ def recv_loop(self):
...
-                    self._poll_single_request(req_id)
+                    is_success = self._poll_single_request(request)
+                    if not is_success:
+                        self._pending_load_reqs.append(request)
                 except Exception as e:
-                    logger.warning(f"Error receiving data for {req_id}: {e}")
+                    self._pending_load_reqs.append(request)
+                    logger.warning(f"Error receiving data for {request_id}: {e}")
 
             time.sleep(0.001)
```
Comment (Sy0307) [inline vllm_omni/distributed/omni_connectors/transfer_adapter/base.py:63]: This sleep step can be triggered? When 'not ready', your code does this: it calls popleft() (which temporarily empties the queue), the poll fails, and then it immediately calls append() to put the same request back. Sleep can not be triggered forever.
